lib/derivation.ml

```diff
@@ -0,0 +1,37 @@
+open Base
+open Fialyzer
+
+let new_var =
+  let i = ref 0 in
+  fun () ->
+    Int.incr i;
+    Printf.sprintf "%s%02d" "v" !i
+
+let rec derive context = function
+  | Types.Val c ->
+     Ok (Types.TyConstant c, Types.Empty)
+  | Types.Var v ->
+     begin match Context.find context v with
+     | Some ty ->
+        Ok(ty, Types.Empty)
+     | None ->
+        Error ("unknown type variable: " ^ v)
+     end
+  | Types.App (f, [arg1]) ->
+     let open Result in
+     derive context f >>= fun (tyf, cf) ->
+     derive context arg1 >>= fun (ty1, c1) ->
+     begin match tyf with
+     | Types.Fun([ty_arg1], ty) ->
+        let beta = new_var() in
+        Ok(Types.TyVar beta, Types.Conj [cf; c1])
```

Comment:
I think this must have subtype constraints (`β ⊆ α ∧ ...`) even if eq relation is already replaced.

Reply:
Thank you, it's fixed.